gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/ExecutionChecksUtil.java

```diff
@@ -0,0 +1,185 @@
+ * not attempt major reorganization as abstractions may change.
+ */
+public final class ExecutionChecksUtil {
+
+  /**
+   * For a given a flowSpec, verifies that an execution is allowed (in case there is an ongoing execution) and the
+   * flowspec can be compiled. If the pre-conditions hold, then a JobExecutionPlan is constructed and returned to the
+   * caller.
+   * @return jobExecutionPlan dag if one can be constructed for the given flowSpec
+   */
+  public static Optional<Dag<JobExecutionPlan>> handleChecksBeforeExecution(
+      SharedFlowMetricsContainer sharedFlowMetricsContainer, SpecCompiler specCompiler, UserQuotaManager quotaManager,
+      Optional<EventSubmitter> eventSubmitter, FlowStatusGenerator flowStatusGenerator, Logger log,
+      boolean flowConcurrencyFlag, FlowSpec flowSpec)
+      throws IOException, InterruptedException {
+    Config flowConfig = flowSpec.getConfig();
+    String flowGroup = flowConfig.getString(ConfigurationKeys.FLOW_GROUP_KEY);
+    String flowName = flowConfig.getString(ConfigurationKeys.FLOW_NAME_KEY);
+    if (!isExecutionPermittedHandler(sharedFlowMetricsContainer, specCompiler, quotaManager, eventSubmitter,
+        flowStatusGenerator, log, flowConcurrencyFlag, flowConfig, flowSpec, flowName, flowGroup)) {
+      return Optional.absent();
+    }
+
+    //Wait for the SpecCompiler to become healthy.
+    specCompiler.awaitHealthy();
+
+    Dag<JobExecutionPlan> jobExecutionPlanDag = specCompiler.compileFlow(flowSpec);
+    Optional<TimingEvent> flowCompilationTimer =
+        eventSubmitter.transform(submitter -> new TimingEvent(submitter, TimingEvent.FlowTimings.FLOW_COMPILED));
+    Map<String, String> flowMetadata = TimingEventUtils.getFlowMetadata(flowSpec);
+
+    if (jobExecutionPlanDag == null || jobExecutionPlanDag.isEmpty()) {
+      populateFlowCompilationFailedEventMessage(eventSubmitter, flowSpec, flowMetadata);
+      return Optional.absent();
+    }
+
+    addFlowExecutionIdIfAbsent(flowMetadata, jobExecutionPlanDag);
+    if (flowCompilationTimer.isPresent()) {
+      flowCompilationTimer.get().stop(flowMetadata);
+    }
+    return Optional.of(jobExecutionPlanDag);
+  }
+
+  /**
+   * Checks if flowSpec disallows concurrent executions, and if so then checks if another instance of the flow is
+   * already running and emits a FLOW FAILED event. Otherwise, this check passes.
+   * @return true if caller can proceed to execute flow, false otherwise
+   * @throws IOException
+   */
+  public static boolean isExecutionPermittedHandler(SharedFlowMetricsContainer sharedFlowMetricsContainer,
+      SpecCompiler specCompiler, UserQuotaManager quotaManager, Optional<EventSubmitter> eventSubmitter,
+      FlowStatusGenerator flowStatusGenerator, Logger log, boolean flowConcurrencyFlag, Config flowConfig, Spec spec,
+      String flowName, String flowGroup) throws IOException {
```

Comment:
I commented elsewhere about lack of uniform ordering between `flowName` and `flowGroup` params... as well as way to sidestep via `FlowId` type